src/main/java/net/minecraftforge/items/PortableInventoryLocation.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+package net.minecraftforge.items;
+
+import net.minecraft.inventory.IInventory;
+import net.minecraft.item.ItemStack;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.IBlockAccess;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public class PortableInventoryLocation
```

Comment:
The idea with this is that most people don't need to know anything about the inventory (like the shulker boxes or normal backpacks). The subclasses are only relevant in cases like IC2 uranium which IMO shouldn't go into any backpacks but the containment box. In such cases it is just as you stated, the parameter has to be `ìnstanceof`-checked and cast to the right type. I can't think of any system that would not require a "default route": Determining whether an item should go into a specific (unknown) type of modded inventory is something that can't be done automatically IMO.
If you have a better idea I'd be glad to hear it.